client/src/pages/EditProfile.js

```diff
@@ -17,59 +18,123 @@ import {
   OptionDiv,
   FormLayout,
   Background,
-} from "../components/EditProfile/EditComponents";
-// dummy data props,context, redux etc
-const editProfile = true;
-
-function EditProfile(props) {
-  // dummy data props,context, redux etc
-  const {
-    firstName,
-    lastName,
-    facebookURL,
-    twitterURL,
-    githubURL,
-    linkedinURL,
-    personalURL,
-    about,
-  } = props.user;
-  const { register, handleSubmit } = useForm();
+} from "components/EditProfile/EditComponents";
+import { UserContext, withUserContext } from "context/UserContext";
+import {
+  fetchUser,
+  fetchUserError,
+  fetchUserSuccess,
+  updateUser,
+  updateUserError,
+  updateUserSuccess,
+} from "hooks/actions/userActions";
+import { getInitials } from "utils/userInfo";
+import { validateURL } from "utils/validators";
 
-  const onSubmit = (data) => {
-    // console.log(data);
-    // make a put/patch request to backend to update users profile information
-  };
+const URLS_CONFIG = {
+  facebook: [
+    "Facebook URL",
+    {
+      pattern: {
+        value: /^[a-zA-Z0-9.]*$/,
+        message:
+          "Invalid entry: only alphanumeric characters and . are allowed",
+      },
+      minLength: {
+        value: 5,
+        message: "Min. length is 5 characters",
+      },
+    },
+    "https://www.facebook.com/",
+  ],
+  linkedin: [
+    "LinkedIn URL",
+    {
+      pattern: {
+        value: /^[a-zA-Z0-9]*$/,
+        message: "Invalid entry: only alphanumeric characters are allowed",
+      },
+    },
+    "https://www.linkedin.com/in/",
+  ],
+  twitter: [
+    "Twitter URL",
+    {
+      pattern: {
+        value: /^[a-zA-Z0-9_]*$/,
+        message:
+          "Invalid entry: only alphanumeric characters and _ are allowed",
+      },
+      maxLength: {
+        value: 15,
+        message: "Max. length is 15 characters",
+      },
+    },
+    "https://twitter.com/",
+  ],
+  github: [
+    "Github URL",
+    {
+      pattern: {
+        value: /^[a-zA-Z0-9_-]*$/,
+        message:
+          "Invalid entry: only alphanumeric characters and _ are allowed",
+      },
+    },
+    "https://github.com/",
+  ],
+  website: [
+    "Personal Website",
+    {
+      validate: (str) => !str || validateURL(str) || "Invalid URL",
+    },
+  ],
+};
+const ABOUT_MAX_LENGTH = 160;
 
-  const labelVariableValue = {
-    // label name, variable name, value
-    "Facebook URL": ["facebookURL", facebookURL],
-    "LinkedIn URL": ["linkedinURL", linkedinURL],
-    "Twitter URL": ["twitterURL", twitterURL],
-    "Github URL": ["githubURL", githubURL],
-    "Personal Website": ["personalURL", personalURL],
-  };
+function EditProfile() {
+  const { userProfileState, userProfileDispatch } = useContext(UserContext);
+  const { errors, formState, register, handleSubmit } = useForm({
+    mode: "change",
+  });
+  const { error, loading, user } = userProfileState;
+  const { firstName = "", lastName = "", urls = {}, about } = user || {};
```

Comment:
The URL object keys are sent with empty strings if not entered in the form:

![image](https://user-images.githubusercontent.com/10546720/84516905-7d94c980-ac9c-11ea-8575-8bfe0d6dee90.png)

I think before sending we should delete any keys in the `url` object sent to the backend which are like `"facebook": ""`

Reply:
Sorry I didn't have recent version. it seems like the bakcend is handling this case now.

Reply:
yeah i think it should be fixed now, this terrible error message could possibly still return from the backend and use these JSON schema and regex rules but if we see those we can hopefully fix it